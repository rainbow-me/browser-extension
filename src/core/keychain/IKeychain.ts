import { Signer } from '@ethersproject/abstract-signer';
import { Mnemonic } from '@ethersproject/hdnode';
import { Wallet } from '@ethersproject/wallet';
import { Address } from 'viem';

export type PrivateKey = string;

export type TWallet = Omit<Wallet, 'address' | 'privateKey'> & {
  address: Address;
  privateKey: PrivateKey;
};

export interface IKeychain {
  type: string;
  serialize(): Promise<unknown>;
  deserialize(options: unknown): Promise<void>;
  addNewAccount(): Promise<Array<Wallet>>;
  addAccountAtIndex(index: number, address: Address): Promise<Address>;
  getAccounts(): Promise<Array<Address>>;
  getSigner(address: Address): Signer;
  exportAccount(address: Address): Promise<PrivateKey>;
  exportKeychain(address: Address): Promise<Mnemonic['phrase']>;
  removeAccount(address: Address): Promise<void>;
}
