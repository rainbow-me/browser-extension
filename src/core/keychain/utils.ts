import { RainbowError, logger } from '~/logger';

import { ahaHttp } from '../network/aha';

const ACCOUNTS_TO_DESERIALIZE = 10;

export const autoDiscoverAccounts = async ({
  deriveWallet,
}: {
  deriveWallet: (index: number) => { address: string };
}): Promise<{ accountsEnabled: number }> => {
  let totalAccountsEnabled = 0;
  let lastAccountsEnabled = ACCOUNTS_TO_DESERIALIZE;
  while (lastAccountsEnabled === ACCOUNTS_TO_DESERIALIZE) {
    // eslint-disable-next-line no-await-in-loop
    const { accountsEnabled } = await autoDiscoverAccountsFromIndex({
      initialIndex: totalAccountsEnabled,
      deriveWallet,
    });
    lastAccountsEnabled = accountsEnabled;
    totalAccountsEnabled += accountsEnabled;
  }
  return {
    accountsEnabled: totalAccountsEnabled === 0 ? 1 : totalAccountsEnabled,
  };
};

export const autoDiscoverAccountsFromIndex = async ({
  initialIndex,
  deriveWallet,
}: {
  initialIndex: number;
  deriveWallet: (index: number) => { address: string };
}): Promise<{ accountsEnabled: number }> => {
  const addresses = Array.from(
    { length: ACCOUNTS_TO_DESERIALIZE },
    (_, i) => initialIndex + i,
  ).map((i) => deriveWallet(i).address);

  try {
    const { data } = await ahaHttp.get(`/?address=${addresses.join(',')}`);
    const addressesHaveBeenUsed = data as {
      data: { addresses: { [key: string]: boolean } };
    };

    const firstNotUsedAddressIndex = addresses.findIndex(
      (address) =>
        !addressesHaveBeenUsed.data.addresses[address?.toLowerCase()],
    );

    return {
      accountsEnabled:
        firstNotUsedAddressIndex === -1
          ? ACCOUNTS_TO_DESERIALIZE
          : firstNotUsedAddressIndex,
    };
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  } catch (error: any) {
    logger.error(new RainbowError(`[aha]: Failed to discover wallets`), {
      message: error.message,
    });
    return { accountsEnabled: 1 };
  }
};
