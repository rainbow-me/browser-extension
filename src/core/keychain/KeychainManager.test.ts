import { isAddress } from '@ethersproject/address';
import { isBytesLike } from '@ethersproject/bytes';
import { Wallet } from '@ethersproject/wallet';
import { beforeAll, expect, test } from 'vitest';

import { delay } from '~/test/utils';

import { KeychainType } from '../types/keychainTypes';

import { PrivateKey } from './IKeychain';
import { keychainManager } from './KeychainManager';

let privateKey = '';
let password = '';

beforeAll(async () => {
  await delay(3000);
});

test('[keychain/KeychainManager] :: should be able to create an HD wallet', async () => {
  await keychainManager.addNewKeychain();
  const accounts = await keychainManager.getAccounts();
  expect(accounts.length).toBe(1);
  expect(isAddress(accounts[0])).toBe(true);
});

test('[keychain/KeychainManager] :: should be able to add an account', async () => {
  let accounts = await keychainManager.getAccounts();
  const hdCreatedKeychain = await keychainManager.getKeychain(accounts[0]);
  await keychainManager.addNewAccount(hdCreatedKeychain);
  accounts = await keychainManager.getAccounts();
  expect(accounts.length).toBe(2);
  expect(isAddress(accounts[1])).toBe(true);
});

test('[keychain/KeychainManager] :: should be able to export a private key for an account', async () => {
  const accounts = await keychainManager.getAccounts();
  privateKey = (await keychainManager.exportAccount(
    accounts[1],
    password,
  )) as PrivateKey;
  expect(isBytesLike(privateKey)).toBe(true);
});

test('[keychain/KeychainManager] :: should be able to remove an account from an HD keychain...', async () => {
  let accounts = await keychainManager.getAccounts();
  await keychainManager.removeAccount(accounts[1]);
  accounts = await keychainManager.getAccounts();
  expect(accounts.length).toBe(1);
});

test('[keychain/KeychainManager] :: should be able to export the seed phrase for an HD wallet', async () => {
  const accounts = await keychainManager.getAccounts();
  const seedPhrase = await keychainManager.exportKeychain(
    accounts[0],
    password,
  );
  expect(seedPhrase.split(' ').length).toBe(12);
});

test('[keychain/KeychainManager] :: should be able to add a read only wallet using an address', async () => {
  await keychainManager.importKeychain({
    type: KeychainType.ReadOnlyKeychain,
    address: '0x70c16D2dB6B00683b29602CBAB72CE0Dcbc243C4',
  });
  const accounts = await keychainManager.getAccounts();
  expect(accounts.length).toBe(2);
  expect(isAddress(accounts[1])).toBe(true);
  expect(accounts[1]).toBe('0x70c16D2dB6B00683b29602CBAB72CE0Dcbc243C4');
});

test('[keychain/KeychainManager] :: should be able to remove an account from a ReadOnly keychain...', async () => {
  let accounts = await keychainManager.getAccounts();
  await keychainManager.removeAccount(accounts[1]);
  accounts = await keychainManager.getAccounts();
  expect(accounts.length).toBe(1);
});

test('[keychain/KeychainManager] :: should be able to import a wallet using a private key', async () => {
  await keychainManager.importKeychain({
    type: KeychainType.KeyPairKeychain,
    privateKey,
  });
  const accounts = await keychainManager.getAccounts();
  expect(accounts.length).toBe(2);
  expect(isAddress(accounts[1])).toBe(true);
});

test('[keychain/KeychainManager] :: should be able to remove an account from a KeyPair keychain...', async () => {
  let accounts = await keychainManager.getAccounts();
  await keychainManager.removeAccount(accounts[1]);
  accounts = await keychainManager.getAccounts();
  expect(accounts.length).toBe(1);
});

test('[keychain/KeychainManager] :: should be able to remove empty keychains', async () => {
  let accounts = await keychainManager.getAccounts();
  await keychainManager.removeAccount(accounts[0]);
  accounts = await keychainManager.getAccounts();
  expect(accounts.length).toBe(0);
  expect(keychainManager.state.keychains.length).toBe(0);
});

test('[keychain/KeychainManager] :: should be able to import a wallet using a seed phrase', async () => {
  let accounts = await keychainManager.getAccounts();
  await keychainManager.importKeychain({
    type: KeychainType.HdKeychain,
    mnemonic:
      'edge caught toy sniff enemy upon genre van tunnel make disorder home',
  });
  accounts = await keychainManager.getAccounts();
  expect(accounts.length).toBe(1);
  expect(isAddress(accounts[0])).toBe(true);
});

test('[keychain/KeychainManager] :: should be able to get the signer of a specific address', async () => {
  const accounts = await keychainManager.getAccounts();
  const signer = (await keychainManager.getSigner(accounts[0])) as Wallet;
  expect(signer._isSigner).toBe(true);
  expect(signer.address).toBe(accounts[0]);
  expect(signer.sendTransaction).toBeDefined();
});

test('[keychain/KeychainManager] :: should be able to update the password of the vault', async () => {
  password = 'newPassword';
  await keychainManager.setPassword(password);
  expect(await keychainManager.verifyPassword(password)).toBe(true);
});

test('[keychain/KeychainManager] :: should be able to lock the vault', async () => {
  await keychainManager.lock();
  expect(keychainManager.state.isUnlocked).toBe(false);
  expect(keychainManager.state.keychains.length).toBe(0);
});

test('[keychain/KeychainManager] :: should be able to unlock the vault', async () => {
  await keychainManager.unlock(password);
  expect(keychainManager.state.isUnlocked).toBe(true);
  expect(keychainManager.state.keychains.length).toBe(1);
});

test('[keychain/KeychainManager] :: should be able to autodiscover accounts when importing a seed phrase', async () => {
  let accounts = await keychainManager.getAccounts();
  expect(keychainManager.state.keychains.length).toBe(1);
  await keychainManager.importKeychain({
    type: KeychainType.HdKeychain,
    mnemonic: 'test test test test test test test test test test test junk',
  });
  expect(keychainManager.state.keychains.length).toBe(2);

  accounts = await keychainManager.getAccounts();
  expect(accounts.length).toBeGreaterThan(2);
  expect(accounts[1]).equal('0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266');
  expect(accounts[2]).toBe('0x70997970C51812dc3A010C7d01b50e0d17dc79C8');

  const privateKey1 = (await keychainManager.exportAccount(
    accounts[1],
    password,
  )) as PrivateKey;
  const privateKey2 = (await keychainManager.exportAccount(
    accounts[2],
    password,
  )) as PrivateKey;

  expect(privateKey1).equal(
    '0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80',
  );
  expect(privateKey2).equal(
    '0x59c6995e998f97a5a0044966f0945389dc9e86dae88c7a8412f4603b6b78690d',
  );
});
