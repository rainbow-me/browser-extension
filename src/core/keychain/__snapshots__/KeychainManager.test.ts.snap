// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`[keychain/KeychainManager] :: should be able to add a read only wallet using an address 1`] = `
{
  "local": {
    "vault": "{\\"data\\":\\"SnoVv/vROdgXlYXUDTsN6GsPgtCrej3ImvQd4H/oBfeaBIV5kYRl2gIUekO0CX5yCIC1cbm1OxI/myn+gqApmUBTN4sYo7l2k3zTnWWRpPCoon1MBMefNxxB+1UlRvxa4OrMM4kRmYn18IbG4L1L0hkwad/3ndP9fzTQ3huEDIptDOeD7hyg1C26wl+78fdbeNR0gHcyJc8CI8p5etelVX0ncYt6oMW4VAqorGatPMzn+8Zt2xBfo9EVq72/zBJib8LdMk18Zgxld/A+nOeh39RCUf3EPSQFuhfaXwLiiwIx9h/Ld3H7CBvcboJ01oef4Q1RbqEbKQ+3gTsS38juxjU7EkzmDNZpCsELLy46LLuzASaSEZsR0A==\\",\\"iv\\":\\"AAECAwQFBgcICQoLDA0ODw==\\",\\"salt\\":\\"AAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh8=\\"}",
  },
  "session": {
    "encryptionKey": "{\\"key_ops\\":[\\"encrypt\\",\\"decrypt\\"],\\"ext\\":true,\\"kty\\":\\"oct\\",\\"k\\":\\"sHyhXbroUejPho-aO3OoHJ7ZCPEl8tYOPvBsd6wRsI8\\",\\"alg\\":\\"A256GCM\\"}",
    "salt": "AAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh8=",
  },
}
`;

exports[`[keychain/KeychainManager] :: should be able to add an account 1`] = `
{
  "local": {
    "vault": "{\\"data\\":\\"SnoVv/vROdgXlYXUDTsN6GsPgtCrej3ImvQd4H/oBfeaBIV5kYRl2gIUekO0CX5yCIC1cbm1OxI/myn+gqApmUBTN4sYo7l2k3zTnWWRpPCoon1MBMefNxxB+1UlRvxa4OrMM4kRmYn18IbG4L1I0hkwad/3ndP9fzTQ3huEDIptDOeD7hyg1C26wl+78fdbeNR0gHcyJc8CI8p5etelVX0ncYt6oMW4VAqorGatPMznl+ZNBlqPv7i2bXfQEF94/E/Qxg==\\",\\"iv\\":\\"AAECAwQFBgcICQoLDA0ODw==\\",\\"salt\\":\\"AAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh8=\\"}",
  },
  "session": {
    "encryptionKey": "{\\"key_ops\\":[\\"encrypt\\",\\"decrypt\\"],\\"ext\\":true,\\"kty\\":\\"oct\\",\\"k\\":\\"sHyhXbroUejPho-aO3OoHJ7ZCPEl8tYOPvBsd6wRsI8\\",\\"alg\\":\\"A256GCM\\"}",
    "salt": "AAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh8=",
  },
}
`;

exports[`[keychain/KeychainManager] :: should be able to autodiscover accounts when importing a seed phrase 1`] = `
{
  "local": {
    "vault": "{\\"data\\":\\"eG/OEE2jXuDMsmRKl10zqlij8ibUWBQIiB4/kMix8XjZFiNjxBZABpcGq5/0tdLsC/paj6WYjNXtf2fcqO6/6Y0nJ5cvTzhJxZTmr8hgXMzEikZimVUd/m7QOjuZojDP3DBI3aCq4etsJ9jl/f29ZBasAwTOeOJ4K+D/HzOFvGRVhZVUEdrpOf3A/ABOxw8l9j2BuGbwAyVsgN0hFSvMnwNlR30PNy9qjaWp66D1ZLkUncIinUO8jRmD3fnWdYD/rkPQKVTG/x7KKGocn/ngclpyXXVweQ7XRZ8ka5mxpHyn13OrLEMLHeUDU9THMXxCZMdc4w4bOR4fEt3qy7SVJJzwtNCwhrqP0J4LpfzwteQxwt8Dyplom/tpW/QpADOar4xmlqjCQ8OrG59RaszGo1QWD5bl6Qljxz5XExlhhuBhwk/mOGvTUpxU9jAJZYotVIV8/WQv/uLCWSrTXxh9GliIDEth3IWAhH+AsUYkc9PhBux30q9j26JoueWd5Q==\\",\\"iv\\":\\"AAECAwQFBgcICQoLDA0ODw==\\",\\"salt\\":\\"AAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh8=\\"}",
  },
  "session": {
    "encryptionKey": "{\\"key_ops\\":[\\"encrypt\\",\\"decrypt\\"],\\"ext\\":true,\\"kty\\":\\"oct\\",\\"k\\":\\"xBwW7Ghxj8m_0T2wIbWvt4dfTLbhDQJ5ynHnJMzlfls\\",\\"alg\\":\\"A256GCM\\"}",
    "salt": "AAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh8=",
  },
}
`;

exports[`[keychain/KeychainManager] :: should be able to create an HD wallet 1`] = `
{
  "local": {
    "vault": "{\\"data\\":\\"SnoVv/vROdgXlYXUDTsN6GsPgtCrej3ImvQd4H/oBfeaBIV5kYRl2gIUekO0CX5yCIC1cbm1OxI/myn+gqApmUBTN4sYo7l2k3zTnWWRpPCoon1MBMefNxxB+1UlRvxa4OrMM4kRmYn18IbG4L1L0hkwad/3ndP9fzTQ3huEDIptDOeD7hyg1C26wl+78fdbeNR0gHcyJc8CI8p5etelVX0ncYt6oMW4VAqorGatPMznl+ZN3wm1OEH98PD9oRWUIgLXtQ==\\",\\"iv\\":\\"AAECAwQFBgcICQoLDA0ODw==\\",\\"salt\\":\\"AAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh8=\\"}",
  },
  "session": {
    "encryptionKey": "{\\"key_ops\\":[\\"encrypt\\",\\"decrypt\\"],\\"ext\\":true,\\"kty\\":\\"oct\\",\\"k\\":\\"sHyhXbroUejPho-aO3OoHJ7ZCPEl8tYOPvBsd6wRsI8\\",\\"alg\\":\\"A256GCM\\"}",
    "salt": "AAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh8=",
  },
}
`;

exports[`[keychain/KeychainManager] :: should be able to export a private key for an account 1`] = `
{
  "local": {
    "vault": "{\\"data\\":\\"SnoVv/vROdgXlYXUDTsN6GsPgtCrej3ImvQd4H/oBfeaBIV5kYRl2gIUekO0CX5yCIC1cbm1OxI/myn+gqApmUBTN4sYo7l2k3zTnWWRpPCoon1MBMefNxxB+1UlRvxa4OrMM4kRmYn18IbG4L1I0hkwad/3ndP9fzTQ3huEDIptDOeD7hyg1C26wl+78fdbeNR0gHcyJc8CI8p5etelVX0ncYt6oMW4VAqorGatPMznl+ZNBlqPv7i2bXfQEF94/E/Qxg==\\",\\"iv\\":\\"AAECAwQFBgcICQoLDA0ODw==\\",\\"salt\\":\\"AAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh8=\\"}",
  },
  "session": {
    "encryptionKey": "{\\"key_ops\\":[\\"encrypt\\",\\"decrypt\\"],\\"ext\\":true,\\"kty\\":\\"oct\\",\\"k\\":\\"sHyhXbroUejPho-aO3OoHJ7ZCPEl8tYOPvBsd6wRsI8\\",\\"alg\\":\\"A256GCM\\"}",
    "salt": "AAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh8=",
  },
}
`;

exports[`[keychain/KeychainManager] :: should be able to export the seed phrase for an HD wallet 1`] = `
{
  "local": {
    "vault": "{\\"data\\":\\"SnoVv/vROdgXlYXUDTsN6GsPgtCrej3ImvQd4H/oBfeaBIV5kYRl2gIUekO0CX5yCIC1cbm1OxI/myn+gqApmUBTN4sYo7l2k3zTnWWRpPCoon1MBMefNxxB+1UlRvxa4OrMM4kRmYn18IbG4L1L0hkwad/3ndP9fzTQ3huEDIptDOeD7hyg1C26wl+78fdbeNR0gHcyJc8CI8p5etelVX0ncYt6oMW4VAqorGatPMzn+8ZtqtrQ1JpYWzhn03mN8uDaxIU=\\",\\"iv\\":\\"AAECAwQFBgcICQoLDA0ODw==\\",\\"salt\\":\\"AAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh8=\\"}",
  },
  "session": {
    "encryptionKey": "{\\"key_ops\\":[\\"encrypt\\",\\"decrypt\\"],\\"ext\\":true,\\"kty\\":\\"oct\\",\\"k\\":\\"sHyhXbroUejPho-aO3OoHJ7ZCPEl8tYOPvBsd6wRsI8\\",\\"alg\\":\\"A256GCM\\"}",
    "salt": "AAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh8=",
  },
}
`;

exports[`[keychain/KeychainManager] :: should be able to get the signer of a specific address 1`] = `
{
  "local": {
    "vault": "{\\"data\\":\\"SnoVv/vROdgXlYXUDSke8X1GjJ+ocTXKm/MXoWfwQueYEtE6hJRxxlZAa1+5CXl5EpLzJbmoKl8y3i/6zbpsjVEdMYtLofhs1nvS03+Ru6Tlt3NaUIOcMR1YvRx1B/dW4vqAa9g1lIv46Y3WscIUn1k0aOu004rzZ37ZoU7EQ4dhHq2D6g+oyzmmxwTypr1JYNoekEUnOY5QYOtzE57+V3Yld5A2+JSdUgWirW29bbLZpv5kkg9f+O4spIV3m+Uux397TbWyIyM8y8ad\\",\\"iv\\":\\"AAECAwQFBgcICQoLDA0ODw==\\",\\"salt\\":\\"AAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh8=\\"}",
  },
  "session": {
    "encryptionKey": "{\\"key_ops\\":[\\"encrypt\\",\\"decrypt\\"],\\"ext\\":true,\\"kty\\":\\"oct\\",\\"k\\":\\"sHyhXbroUejPho-aO3OoHJ7ZCPEl8tYOPvBsd6wRsI8\\",\\"alg\\":\\"A256GCM\\"}",
    "salt": "AAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh8=",
  },
}
`;

exports[`[keychain/KeychainManager] :: should be able to import a wallet using a private key 1`] = `
{
  "local": {
    "vault": "{\\"data\\":\\"SnoVv/vROdgXlYXUDTsN6GsPgtCrej3ImvQd4H/oBfeaBIV5kYRl2gIUekO0CX5yCIC1cbm1OxI/myn+gqApmUBTN4sYo7l2k3zTnWWRpPCoon1MBMefNxxB+1UlRvxa4OrMM4kRmYn18IbG4L1L0hkwad/3ndP9fzTQ3huEDIptDOeD7hyg1C26wl+78fdbeNR0gHcyJc8CI8p5etelVX0ncYt6oMW4VAqorGatPMzn+8Zt2xBfsscYr7m42ns9NNCfJ01nYgNCc61Fn8Co1tocW6iXMXMFuGCsJ3bj33FN8m6bLSutXxH6YpY8lZbeqQtENbt9fVy1+mNOleG93WFrRRy4SokY9iQT8x2T9JVX53WogRHuRJ8lYN4QhTtlkUvwIhrWDVIh42+2AeObVxse\\",\\"iv\\":\\"AAECAwQFBgcICQoLDA0ODw==\\",\\"salt\\":\\"AAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh8=\\"}",
  },
  "session": {
    "encryptionKey": "{\\"key_ops\\":[\\"encrypt\\",\\"decrypt\\"],\\"ext\\":true,\\"kty\\":\\"oct\\",\\"k\\":\\"sHyhXbroUejPho-aO3OoHJ7ZCPEl8tYOPvBsd6wRsI8\\",\\"alg\\":\\"A256GCM\\"}",
    "salt": "AAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh8=",
  },
}
`;

exports[`[keychain/KeychainManager] :: should be able to import a wallet using a seed phrase 1`] = `
{
  "local": {
    "vault": "{\\"data\\":\\"SnoVv/vROdgXlYXUDSke8X1GjJ+ocTXKm/MXoWfwQueYEtE6hJRxxlZAa1+5CXl5EpLzJbmoKl8y3i/6zbpsjVEdMYtLofhs1nvS03+Ru6Tlt3NaUIOcMR1YvRx1B/dW4vqAa9g1lIv46Y3WscIUn1k0aOu004rzZ37ZoU7EQ4dhHq2D6g+oyzmmxwTypr1JYNoekEUnOY5QYOtzE57+V3Yld5A2+JSdUgWirW29bbLZpv5kkg9f+O4spIV3m+Uux397TbWyIyM8y8ad\\",\\"iv\\":\\"AAECAwQFBgcICQoLDA0ODw==\\",\\"salt\\":\\"AAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh8=\\"}",
  },
  "session": {
    "encryptionKey": "{\\"key_ops\\":[\\"encrypt\\",\\"decrypt\\"],\\"ext\\":true,\\"kty\\":\\"oct\\",\\"k\\":\\"sHyhXbroUejPho-aO3OoHJ7ZCPEl8tYOPvBsd6wRsI8\\",\\"alg\\":\\"A256GCM\\"}",
    "salt": "AAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh8=",
  },
}
`;

exports[`[keychain/KeychainManager] :: should be able to lock the vault 1`] = `
{
  "local": {
    "vault": "{\\"data\\":\\"eG/OEE2jXuDMsmRKl10zqlij8ibUWBQIiB4/kMix8XjZFiNjxBZABpcGq5/0tdLsC/paj6WYjNXtf2fcqO6/6Y0nJ5cvTzhJxZTmr8hgXMzEikZimVUd/m7QOjuZojDP3DBI3aCq4etsJ9jl/f29ZBasAwTOeOJ4K+D/HzOFvGRVhZVUEdrpOf3A/ABOxw8l9j2BuGbwAyVsgN0hFSvMnwNlR30PNy9qjaWp66D1ZLkUncIinUO8jRmD3Yghhd+WwsErHajIfyOTtxW2\\",\\"iv\\":\\"AAECAwQFBgcICQoLDA0ODw==\\",\\"salt\\":\\"AAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh8=\\"}",
  },
  "session": {
    "encryptionKey": null,
    "salt": null,
  },
}
`;

exports[`[keychain/KeychainManager] :: should be able to remove an account from a KeyPair keychain... 1`] = `
{
  "local": {
    "vault": "{\\"data\\":\\"SnoVv/vROdgXlYXUDTsN6GsPgtCrej3ImvQd4H/oBfeaBIV5kYRl2gIUekO0CX5yCIC1cbm1OxI/myn+gqApmUBTN4sYo7l2k3zTnWWRpPCoon1MBMefNxxB+1UlRvxa4OrMM4kRmYn18IbG4L1L0hkwad/3ndP9fzTQ3huEDIptDOeD7hyg1C26wl+78fdbeNR0gHcyJc8CI8p5etelVX0ncYt6oMW4VAqorGatPMzn+8ZtqtrQ1JpYWzhn03mN8uDaxIU=\\",\\"iv\\":\\"AAECAwQFBgcICQoLDA0ODw==\\",\\"salt\\":\\"AAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh8=\\"}",
  },
  "session": {
    "encryptionKey": "{\\"key_ops\\":[\\"encrypt\\",\\"decrypt\\"],\\"ext\\":true,\\"kty\\":\\"oct\\",\\"k\\":\\"sHyhXbroUejPho-aO3OoHJ7ZCPEl8tYOPvBsd6wRsI8\\",\\"alg\\":\\"A256GCM\\"}",
    "salt": "AAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh8=",
  },
}
`;

exports[`[keychain/KeychainManager] :: should be able to remove an account from a ReadOnly keychain... 1`] = `
{
  "local": {
    "vault": "{\\"data\\":\\"SnoVv/vROdgXlYXUDTsN6GsPgtCrej3ImvQd4H/oBfeaBIV5kYRl2gIUekO0CX5yCIC1cbm1OxI/myn+gqApmUBTN4sYo7l2k3zTnWWRpPCoon1MBMefNxxB+1UlRvxa4OrMM4kRmYn18IbG4L1L0hkwad/3ndP9fzTQ3huEDIptDOeD7hyg1C26wl+78fdbeNR0gHcyJc8CI8p5etelVX0ncYt6oMW4VAqorGatPMzn+8ZtqtrQ1JpYWzhn03mN8uDaxIU=\\",\\"iv\\":\\"AAECAwQFBgcICQoLDA0ODw==\\",\\"salt\\":\\"AAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh8=\\"}",
  },
  "session": {
    "encryptionKey": "{\\"key_ops\\":[\\"encrypt\\",\\"decrypt\\"],\\"ext\\":true,\\"kty\\":\\"oct\\",\\"k\\":\\"sHyhXbroUejPho-aO3OoHJ7ZCPEl8tYOPvBsd6wRsI8\\",\\"alg\\":\\"A256GCM\\"}",
    "salt": "AAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh8=",
  },
}
`;

exports[`[keychain/KeychainManager] :: should be able to remove an account from an HD keychain... 1`] = `
{
  "local": {
    "vault": "{\\"data\\":\\"SnoVv/vROdgXlYXUDTsN6GsPgtCrej3ImvQd4H/oBfeaBIV5kYRl2gIUekO0CX5yCIC1cbm1OxI/myn+gqApmUBTN4sYo7l2k3zTnWWRpPCoon1MBMefNxxB+1UlRvxa4OrMM4kRmYn18IbG4L1L0hkwad/3ndP9fzTQ3huEDIptDOeD7hyg1C26wl+78fdbeNR0gHcyJc8CI8p5etelVX0ncYt6oMW4VAqorGatPMzn+8ZtqtrQ1JpYWzhn03mN8uDaxIU=\\",\\"iv\\":\\"AAECAwQFBgcICQoLDA0ODw==\\",\\"salt\\":\\"AAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh8=\\"}",
  },
  "session": {
    "encryptionKey": "{\\"key_ops\\":[\\"encrypt\\",\\"decrypt\\"],\\"ext\\":true,\\"kty\\":\\"oct\\",\\"k\\":\\"sHyhXbroUejPho-aO3OoHJ7ZCPEl8tYOPvBsd6wRsI8\\",\\"alg\\":\\"A256GCM\\"}",
    "salt": "AAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh8=",
  },
}
`;

exports[`[keychain/KeychainManager] :: should be able to remove empty keychains 1`] = `
{
  "local": {
    "vault": "",
  },
  "session": {
    "encryptionKey": "{\\"key_ops\\":[\\"encrypt\\",\\"decrypt\\"],\\"ext\\":true,\\"kty\\":\\"oct\\",\\"k\\":\\"sHyhXbroUejPho-aO3OoHJ7ZCPEl8tYOPvBsd6wRsI8\\",\\"alg\\":\\"A256GCM\\"}",
    "salt": "AAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh8=",
  },
}
`;

exports[`[keychain/KeychainManager] :: should be able to unlock the vault 1`] = `
{
  "local": {
    "vault": "{\\"data\\":\\"eG/OEE2jXuDMsmRKl10zqlij8ibUWBQIiB4/kMix8XjZFiNjxBZABpcGq5/0tdLsC/paj6WYjNXtf2fcqO6/6Y0nJ5cvTzhJxZTmr8hgXMzEikZimVUd/m7QOjuZojDP3DBI3aCq4etsJ9jl/f29ZBasAwTOeOJ4K+D/HzOFvGRVhZVUEdrpOf3A/ABOxw8l9j2BuGbwAyVsgN0hFSvMnwNlR30PNy9qjaWp66D1ZLkUncIinUO8jRmD3Yghhd+WwsErHajIfyOTtxW2\\",\\"iv\\":\\"AAECAwQFBgcICQoLDA0ODw==\\",\\"salt\\":\\"AAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh8=\\"}",
  },
  "session": {
    "encryptionKey": "{\\"key_ops\\":[\\"encrypt\\",\\"decrypt\\"],\\"ext\\":true,\\"kty\\":\\"oct\\",\\"k\\":\\"xBwW7Ghxj8m_0T2wIbWvt4dfTLbhDQJ5ynHnJMzlfls\\",\\"alg\\":\\"A256GCM\\"}",
    "salt": "AAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh8=",
  },
}
`;

exports[`[keychain/KeychainManager] :: should be able to update the password of the vault 1`] = `
{
  "local": {
    "vault": "{\\"data\\":\\"eG/OEE2jXuDMsmRKl10zqlij8ibUWBQIiB4/kMix8XjZFiNjxBZABpcGq5/0tdLsC/paj6WYjNXtf2fcqO6/6Y0nJ5cvTzhJxZTmr8hgXMzEikZimVUd/m7QOjuZojDP3DBI3aCq4etsJ9jl/f29ZBasAwTOeOJ4K+D/HzOFvGRVhZVUEdrpOf3A/ABOxw8l9j2BuGbwAyVsgN0hFSvMnwNlR30PNy9qjaWp66D1ZLkUncIinUO8jRmD3Yghhd+WwsErHajIfyOTtxW2\\",\\"iv\\":\\"AAECAwQFBgcICQoLDA0ODw==\\",\\"salt\\":\\"AAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh8=\\"}",
  },
  "session": {
    "encryptionKey": "{\\"key_ops\\":[\\"encrypt\\",\\"decrypt\\"],\\"ext\\":true,\\"kty\\":\\"oct\\",\\"k\\":\\"xBwW7Ghxj8m_0T2wIbWvt4dfTLbhDQJ5ynHnJMzlfls\\",\\"alg\\":\\"A256GCM\\"}",
    "salt": "AAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh8=",
  },
}
`;
